### PR TITLE
PreExecution non-repudiation implementation

### PR DESCRIPTION
**Pull request by tdimitrov**
This PR implements PreExecution result verification by non-primary replicas (aka PreExecution non-repudiation).

What it does:
1. Adds flag to the PreProcessor indicating if the feature should be enabled or not (feature flag).
2. Adds new parameter to ClientRequestMessage, which allows allocating extra buffer space by MessageBase class. This buffer can be used by messages inheriting from ClientRequestMessage to store additional fields (see next point).
3. Adds new PreExecution message - PreProcessedResult message. It is carries the result from the PreExecution and the signatures of all f+1 replicas which has preexecuted the concrete request. This message is implemented by inheriting from ClientRequestMessage. The feature described in point 2 is used to allocate additional buffer space for the signatures.
4. Result verification logic handling.

How the result verification logic changes the PreExecution flow?
================================================
The flow remains unchanged up to receiving a PreProcessReply message on the primary. Now instead of adding a ClientRequestMsg to its internal queue the primary uses PreProcessedResult for this purpose. A clear benefit from this is responsibility separation - ClientRequest messages are used for client requests and PreProcessedResult - for pre-executed requests. All hash signatures from the non-primary replicas are added in the PreProcessedResult message. All its other fields are 1:1 compared to ClientRequest.

On finalise preprocessing phase (finalizePreprocessing method) the Primary bundles some ClientRequest and PreProcessedResult messages in a PrePrepare message. There is another benefit from the inheritance here - PrePrepare handling logic doesn't care about the new message. From its point of view it is just another ClientRequest message.

When a non-primary receives a PrePrepare message it unbundles all ClientRequest and PreProcessedResult messages from it. They are discriminated by the message type in the message header. If the non-primary find a PreProcessedResult message it verifies the validity of all signatures in it and additionally if f+1 signatures are assigned. If both requirements are satisfied the message is processed as usual.

However if a non-primary finds a PreProcessedResult message with invalid signature or less than f+1 signatures it assumes that the primary is malicious and triggers a view change. New reason is added to ReplicaAsksToLeaveViewMsg -  PrimarySentBadPreProcessedResult.

### CODE REVIEW

**Comment by tdimitrov** [comment]
@yuliasherman @arc-vmware this is still work in progress. I suggest to delay you review with a few days.

File: bftengine/src/preprocessor/RequestProcessingState.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/RequestProcessingState.cpp:109]: Didn't you lose "++"?

File: bftengine/src/preprocessor/RequestProcessingState.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/RequestProcessingState.cpp:205]: Actually, none of RequestProcessingState functions is thread-safe. Maybe worth adding a general comment that they are guarded by RequestState::mutex from PreProcessor.

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:3]: Nit: completely new file: 2021

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:26]: Consider passing a string instead.

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:90]: This should be verified in validate() function, which looks not implemented.

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:75]: We should know ahead how long is the signatures list. The signature size is also constant.

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.hpp
Comment (yuliasherman) [inline bftengine/src/preprocessor/messages/PreProcessedResultMsg.hpp:3]: ->2021

**Review by yuliasherman — COMMENTED** [review]
Great job!

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:691]: RVO kicks in here so no copy is made. We can't capture the return by reference, because it's a local variable inside deserializeResultSignatureList().

Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:691]: OK, I've missed that we return a local variable.

File: bftengine/src/preprocessor/RequestProcessingState.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/RequestProcessingState.cpp:109]: Sorry, but I don't follow. Which value should be incremented?

Comment (yuliasherman) [reply bftengine/src/preprocessor/RequestProcessingState.cpp:109]: Forget about it, it's OK.

File: bftengine/include/bftengine/ClientMsgs.hpp
Comment (arc-vmware) [inline bftengine/include/bftengine/ClientMsgs.hpp:43]: Will this make Client request message backward incompatible. We should check if this addition effects serialisation and deserialisation of this message.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:703]: We should create the hashes in parallel. Its done in PrePrepareMsg creation.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:706]: its better to pass '\0' instead of 0 as the second parameter is a char.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:707]: we should static_cast instead of the c-style cast.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:708]: if replica himself gets his own message, then will there be any chance of changes? If no then we can just avoid checking the signature for same replica.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:728]: The validate() method of PrePrepareMsg is doing the same. We can call that function here.

File: bftengine/src/bftengine/ReplicaImp.hpp
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.hpp:487]: As this is validate [read only] method, should msg be const& ?

File: bftengine/src/bftengine/messages/ClientRequestMsg.hpp
Comment (arc-vmware) [inline bftengine/src/bftengine/messages/ClientRequestMsg.hpp:80]: nit: If we need only length then also this will result into buffer pointer calculation.

File: bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp
Comment (arc-vmware) [inline bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp:90]: nit: [very minor comment] void PreProcessReplyMsg::getResultHashSignature(std::vector<char> & result_hash) -- This will avoid result vector and copy ctors.

File: bftengine/src/preprocessor/PreProcessor.cpp
Comment (arc-vmware) [inline bftengine/src/preprocessor/PreProcessor.cpp:814]: nit: Nothing to change, just as assertion : if we keep this ClientRequestMsg then also it should be fine, since preprocess result msg is a type of ClientRequestMsg.

**Review by arc-vmware — COMMENTED** [review]
Great work, enjoyed and learned quite a lot. I have few minor comments, feel free to defer if they are not relevant.

File: bftengine/src/preprocessor/PreProcessor.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/PreProcessor.cpp:839]: I agree, will do it for sigsBuf too

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp:75]: I agree that we should know that there will be f signatures and its size is constant, but I have a reason to do it in such a way.

I put the signature lengths so that the message encoding/decoding is decoupled from the SIgManager. I think it's better to pass value sizes in the message unless they are a hard constant. In this case I don't consider them hard constant, because if we change the hashing algorithm this size will change. Replicas will easily figure this out, but the message will be unparsable outside Concord.

My motivation about the list size is similar - we can do for loop from 0 to F, but I wanted to implement more independant message format. I could have put the list size in the message too, but it seemed redundant for me.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:45]: It was painful, but I did it ;)

File: bftengine/include/bftengine/ClientMsgs.hpp
Comment (tdimitrov) [reply bftengine/include/bftengine/ClientMsgs.hpp:43]: Actually this change doesn't affect the ClientRequestMsg. It's a mechanism to allocate extra buffer space if you are inheriting from the message (as it's done for PreProcess(ed)Result message).

The ClientRequestMsg message itself is not affected by this change because the extraDataLength is always set to 0 for it.

I also had concerns for serialisation/deserialisation so I added additional test cases in the message unit test. Check my changes in ClientRequestMsg_test.cpp/

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:703]: Good point, I'll do it.

File: bftengine/src/preprocessor/PreProcessor.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/PreProcessor.cpp]: Initially we will keep both logics and switch between them via configuration. But later we have to remove the old approach (with ClientRequestMsg reuse), because it is not optimal.

File: bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp:90]: Actually no copy happens here too. RVO is applied for return-by-value vector. 

This function is called in RequestProcessingState::handlePreProcessReplyMsg(), as a parameter for emplace_back() which accepts universal reference. This is one of the cases where copy elision is applied so we are fine I think.

File: bftengine/src/bftengine/messages/ClientRequestMsg.hpp
Comment (tdimitrov) [reply bftengine/src/bftengine/messages/ClientRequestMsg.hpp:80]: True, but we usually need both. Even if the pointer is not used I think it's a small price to pay.

File: bftengine/src/bftengine/ReplicaImp.hpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.hpp]: Good point. I tried this while writing the function. There is a call to msg.getResultSignaturesBuf() in it:
`const auto [buf, buf_len] = msg.getResultSignaturesBuf();`
getResultSignaturesBuf returns pointer so it's not correct to be const. This was the only reason to  get non-const reference in validatePreProcessedResultSignatures().

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:702]: My idea about PreProcess(ed)ResultMsg is to be just a carrier for the signatures without actually validating them.

Initially I wanted to have a validate() method which does the signature checks. However this validate is called very early in the message processing code and if it fails the message is discarded. This doesn't work for us because we want to initiate a view change if signature verification fails. 

After considering pros and cons I decided that it's better to handle this in ReplicaImp because (1) it makes more sense to have a view change triggering logic there and (2) I didn't want to couple PreProcess(ed)ResultMsg with SigManager too much. I think this will bite us in the future when we decide to rework the message encoding/decoding code.

Does this make sense?

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:725]: Please check my previous comment: https://github.com/vmware/concord-bft/pull/1639#discussion_r665335187

File: bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp
Comment (arc-vmware) [reply bftengine/src/preprocessor/messages/PreProcessReplyMsg.cpp:90]: yeah, i missed the RVO. this is absolutely fine. Thanks @tdimitrov

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp:707]: We can't static_cast from unsigned to signed, so I'll replace it with reinterpret_cast.

File: bftengine/include/bftengine/ClientMsgs.hpp
Comment (arc-vmware) [reply bftengine/include/bftengine/ClientMsgs.hpp:43]: Checked the unit test, I think this should work.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:702]: We all the time trying to offload the ReplicaImp code from a non-essential functionally, but the functions you added break this approach. I think it's less evil to have PreProcessResultMsg be aware of SigManager than adding all this functionality to ReplicaImp. In case a VC is required, the corresponding PreProcessResultMsg function needs just to return a flag.

Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Yes, message validation can be separated from ViewChange triggering. I've fixed this.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp]: This is also fixed.

File: bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp
Comment (tdimitrov) [reply bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp]: This check is "deferred" due to the VC trigger we talked about. This is the only reason not to put it in validate().

Is this acceptable after the validation code is removed from ReplicaImp?

Comment (yuliasherman) [reply bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp]: Yes, thanks.

File: bftengine/src/bftengine/ReplicaImp.cpp
Comment (tdimitrov) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Actually the hash is calculated only once and should be outside the for loop. Good catch!

**Comment by tdimitrov** [comment]
Thanks for the feedback @yuliasherman @arc-vmware . I have addressed your comments. There are a bunch of commits which I will squash before merging.
Could you please approve the PR if you are satisfied with the changes?

**Review by yuliasherman — APPROVED** [review]
LGTM, 10x!

**Review by arc-vmware — APPROVED** [review]
Awesome, great feature getting in for concord.

**Pull request merged**
